### PR TITLE
Large Cardboard Box Decals/Sprites

### PR DESCRIPTION
**Pull request by FlimFlamm**
This PR adds a new "cardboard_boxes.dmi" file (containing role based box decal sprites and old default box sprites) and modifies cardboardbox.dm in order to update the new sprite path, and to create a "re-coloring mechanic" which gives an option to choose a decal when an open cardboard box is attacked by a pen.

**Edit 7:** 10 Commits later and this addition is much more polished: Code simplified, user selection list text formatted, error checks and error text feedback added, null input handle added, re-sprite item made to be **pen** to align with re-spriting convention.

boxespaint

:cl: FlimFlamm
rscadd: New decals for boxes. Just apply a pen to change them (while they're open)!
/:cl:

### CODE REVIEW

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 1]
```diff
@@ -52,7 +53,8 @@
 	I.alpha = 0
 	animate(I, pixel_z = 32, alpha = 255, time = 5, easing = ELASTIC_EASING)
 
-/obj/structure/closet/cardboard/attackby(obj/item/weapon/W as obj, mob/user as mob, params)
+
+obj/structure/closet/cardboard/attackby(obj/item/weapon/W as obj, mob/user as mob, params)
```
Comment (Alexshreds) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Uhh, why did you remove the /   ?

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: For whatever reason, I was getting compile issues on that line IIRC

**Comment by Nienhaus** [comment]
Huh.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 1 above)
Comment (tigercat2000) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Please readd the `/`.

**Comment by Iamgoofball** [comment]
:-1: where's my Paz box

**Comment by FlimFlamm** [comment]
@Iamgoofball PAZ? If I knew what that was, I would totally sprite it in!

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 1 above)
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: @tigercat2000 readded!

**Comment by Iamgoofball** [comment]
>>>>>editing the cardboard box
>>>>>doesn't understand the MGS meme it comes from

**Comment by FlimFlamm** [comment]
@Iamgoofball Yea I looked it up... heh.... That's supposed to be the default version of the existing box right? 

It's still in there.

**Comment by FlimFlamm** [comment]
@Nienhaus You're probably wondering :"why?"... "just... why?"... I had been steadily trying to play as a shadow person and I found the box was invaluable for doing so. I figured if i had a box that represented my occupation, it would be overall a much more thoroughly enjoyable experience... After naively thinking that shadow people should become the next karma race, I made the sprites and got carried away. After sharing them it turns out people seem to really like them (I know I do) and since sprites are allegedly very cheap resource wise, I eventually got around to learning to implement them (this was my first paracode project).

These bring nothing but fluff, but as far as fluff goes they sure are nice to look at, hopefully at an acceptably cheap resource cost.

**Comment by Nienhaus** [comment]
I was more saying huh to the edits of my box sprite, which looks weird when it's opened so I wasn't fully sure why it's still used.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /obj/structure/closet/cardboard
 	name = "large cardboard box"
 	desc = "Just a box..."
+	icon = 'icons/obj/boxes.dmi'
```
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: I can't actually comment directly on the icon file, but it's named `Boxes.dmi`, and it's generally preferred to not have filenames capitalized like that.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 3]
```diff
@@ -62,3 +64,16 @@
 			for(var/mob/M in viewers(src))
 				M.show_message("<span class='notice'>\The [src] has been cut apart by [user] with \the [WC].</span>", 3, "You hear cutting.", 2)
 			qdel(src)
+			return
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons)) /////addendum by flimflamm
+			var/decalselection = input("Please select a decal") in list("atmos", "bar", "barber", "blueshield",	"brigphysician", "captain",
+				"cargo", "ce",	"caplain",	"chef", "chemist", "civilian", "clown", "cmo", "coroner", "detective", "engineering", "genetics", "hop",
+				"hos", "hydroponics", "ia", "janitor",	"magistrate", "mechanic", "medical", "mime", "mining", "ntrep", "paramedic", "podpilot",
+				"prisoner",	"rd", "security", "syndicate", "therapist", "virology", "warden", "xenobiology")
```
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Heh, good eye, thanks. Going to get that updated lickity split.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: rather than making an associative list, I'm going to rename the sprites themselves such that they have reasonably pretty names. The whole idea being that an extra list or an array of any kind is needless extra code.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 4]
```diff
@@ -62,3 +64,16 @@
...
+
+			icon_opened = addtext("cardboard_open_", decalselection)
+			icon_closed = addtext("cardboard_", decalselection)
```
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: I'll see what I can do about the null input, I was also thinking of one for when people move too far away as well.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 3 above)
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Making the names look decent is hardly "needless".

I guess you could probably get away with using normal names in this list, and just feed them through `ckey()` to simplify them to appropriate state names... though it sounds like you're going to try putting full names into the icon state names. Not sure if that will work out.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 4 above)
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Oh, right, you need a sanity check for adjacency after the input. I always forget about those.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: @ Krausus So I'm kind of wondering how i ought to be handling the null input. As far as i know trying to concatenate a string with a null var will treat the null var as blank text? (if so I don't see this causing an issue but i don't know how the list and subsequent code terminate themselves in teh case of a disconnection caused null input.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: @Krausus, also, is there a standard distance proc check/set of sanity checks i should be using?

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 5]
```diff
@@ -62,3 +64,17 @@
 			for(var/mob/M in viewers(src))
 				M.show_message("<span class='notice'>\The [src] has been cut apart by [user] with \the [WC].</span>", 3, "You hear cutting.", 2)
 			qdel(src)
+			return
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons)) /////addendum by flimflamm
+			var/decalselection = input("Please select a decal") in list( "cancel", "atmospherics", "bartender", "barber", "blueshield",	"brig_physician", "captain",
```
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: You don't need a cancel option in the list. If you change your code from this:
`input() in list()`
to this:
`input() as null|anything in list()`
the input box will get a "Cancel" button, and pressing that will return null.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 6]
```diff
@@ -62,3 +64,17 @@
...
+				"cargo", "chief_engineer",	"chaplain",	"chef", "chemist", "civilian", "clown", "cmo", "coroner", "detective", "engineering", "genetics", "hop",
+				"hos", "hydroponics", "internal_affairs_agent", "janitor",	"magistrate", "mechanic", "medical", "mime", "mining", "nt_representative", "paramedic", "pod_pilot",
+				"prisoner",	"research_director", "security", "syndicate", "therapist", "virology", "warden", "xenobiology")
+			if(decalselection == "cancel")
+				return
```
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Once you're canceling as null, the `if` for this return can just be `if(!decalselection)`.

> also, is there a standard distance check proc/set of sanity checks i should be using?

The standard adjacency check is `sometThing.Adjacent(otherThing)`. You'll want to make sure the box is adjacent to the user, and the crayons are still in their active hand (`W == user.get_active_hand()`). You'll also want to check `user.incapacitated()` - if true, the user is no longer in any condition to finish the task.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: This is really very helpful, thanks so much! I'm going to work on getting these implemented!

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: @Krausus I think i successfully implemented adjacency check, item in hands check, incapacitation check, and also made it consume the crayon box upon completion. The only other potential thing that is left to add would be an associative list, but I already spruced up the sprite names to where they are quite legible as is.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 7]
```diff
@@ -62,3 +64,22 @@
 			for(var/mob/M in viewers(src))
 				M.show_message("<span class='notice'>\The [src] has been cut apart by [user] with \the [WC].</span>", 3, "You hear cutting.", 2)
 			qdel(src)
+			return
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons)) /////addendum by flimflamm
+			var/decalselection = input("Please select a decal") as null|anything in list("atmospherics", "bartender", "barber", "blueshield",	"brig_physician", "captain",
+				"cargo", "chief_engineer",	"chaplain",	"chef", "chemist", "civilian", "clown", "cmo", "coroner", "detective", "engineering", "genetics", "hop",
+				"hos", "hydroponics", "internal_affairs_agent", "janitor",	"magistrate", "mechanic", "medical", "mime", "mining", "nt_representative", "paramedic", "pod_pilot",
+				"prisoner",	"research_director", "security", "syndicate", "therapist", "virology", "warden", "xenobiology")
```
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: You should *really* handle this list in a manner that uses capital letters and spaces. There is no good reason to leave it in this state.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: I think if capitalize the list as it currently exists and can figure out a way to have any spaces replaced with an underscore, then it should achieve this without needing a second or an associative list. I'll get on this.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 8]
```diff
@@ -62,3 +64,22 @@
...
+			if(user.incapacitated())
+				return
+			if(W == user.get_active_hand() & src.Adjacent(usr))
+				icon_opened = ("cardboard_open_"+decalselection)
+				icon_closed = ("cardboard_"+decalselection)
+				update_icon() // a proc declared in the closets parent file used to update opened/closed sprites on normal closets
+				qdel(W)
+				return
+			return
+	return
```
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: So I take this to mean that returns are implied at the end of a code file?

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 9]
```diff
@@ -62,3 +64,22 @@
...
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons)) /////addendum by flimflamm
+			var/decalselection = input("Please select a decal") as null|anything in list("atmospherics", "bartender", "barber", "blueshield",	"brig_physician", "captain",
+				"cargo", "chief_engineer",	"chaplain",	"chef", "chemist", "civilian", "clown", "cmo", "coroner", "detective", "engineering", "genetics", "hop",
+				"hos", "hydroponics", "internal_affairs_agent", "janitor",	"magistrate", "mechanic", "medical", "mime", "mining", "nt_representative", "paramedic", "pod_pilot",
+				"prisoner",	"research_director", "security", "syndicate", "therapist", "virology", "warden", "xenobiology")
+			if(!decalselection)
+				return
+			if(user.incapacitated())
+				return
+			if(W == user.get_active_hand() & src.Adjacent(usr))
```
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: I was unsure about this, thanks!

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 7 above)
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: You'll probably want to use the `lowertext` and `replacetext` procs.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 8 above)
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: At the end of every proc, yes.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 7 above)
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Actually I've been trying to get the replace text proc to work right on this for over an hour now hehe so far im at  **decalselection = replacetext(decalselection, "\\s", "\\x5F")** but for whatever reason it's not returning the correct result. lowertext works fine though.

Any idea on what I'm doing wrong?

Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Use a space and an underscore. `" ", "_")`

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Yep that'll do it. When I first started on the replacetext proc i had not yet found the lowertext proc, and so when I tried initially with an actual space and an actual underscore it wouldn't work, which sent me spiraling to find alternative identifiers (thinking space or underscore was not recognizable).

Now that I've got this complete, I think the one remaining finishing touch will be a check to make sure that it is a full box of crayons.

**Comment by Davdichan** [comment]
I forget, is it possible to put posters on boxes?

**Comment by FlimFlamm** [comment]
@davidchan I don't think so, but I've never tried!

**Comment by FlimFlamm** [comment]
Assuming I've no errors in the most recent commit, at this point I think it is complete!

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 10]
```diff
@@ -62,3 +64,22 @@
 			for(var/mob/M in viewers(src))
 				M.show_message("<span class='notice'>\The [src] has been cut apart by [user] with \the [WC].</span>", 3, "You hear cutting.", 2)
 			qdel(src)
+			return
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons)) /////addendum by flimflamm
```
Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: @Krausus the most recent commit as of this comment performs checks to make sure the box is filled with crayons rather than photocopys of someones butt, or worst, nothing! It doesn't force you to have one of each basic color crayon, but that should not be an issue.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 11]
```diff
@@ -62,3 +64,34 @@
...
 				M.show_message("<span class='notice'>\The [src] has been cut apart by [user] with \the [WC].</span>", 3, "You hear cutting.", 2)
 			qdel(src)
+			return
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons))
+			var/list/contents = W.contents
+			var/crayoncount = 0
+			for(var/C in contents)
+				if(istype(C, /obj/item/toy/crayon))
+					crayoncount += 1
+				continue
```
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: This `continue` statement isn't necessary. The loop automatically continues at the end of its block unless you `break` it.

You don't really need the loop, though. You're dealing with a container that's only allowed to contain crayons, and want to know if it's full. You can simply compare the quantity of contents to its maximum capacity:
`if(W.contents.len == W.storage_slots)`

As an added bonus, you'll no longer need to assume a box of crayons holds exactly 6 of them.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 12]
```diff
@@ -62,3 +64,34 @@
...
+			if(crayoncount == 6)
+				var/decalselection = input("Please select a decal") as null|anything in list("Atmospherics", "Bartender", "Barber", "Blueshield",	"Brig Physician", "Captain",
+					"Cargo", "Chief Engineer",	"Chaplain",	"Chef", "Chemist", "Civilian", "Clown", "CMO", "Coroner", "Detective", "Engineering", "Genetics", "HOP",
+					"HOS", "Hydroponics", "Internal Affairs Agent", "Janitor",	"Magistrate", "Mechanic", "Medical", "Mime", "Mining", "NT Representative", "Paramedic", "Pod Pilot",
+					"Prisoner",	"Research Director", "Security", "Syndicate", "Therapist", "Virology", "Warden", "Xenobiology")
+				if(!decalselection)
+					return
+				if(user.incapacitated())
+					return
+				if(W == user.get_active_hand() && Adjacent(usr))
```
Comment (Krausus) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: Your code doesn't end up getting *too* deep, but whenever possible, you want to bail out instead of adding another indentation level. Just like the checks immediately above this, you can format this with a return:
```
if(!(W == user.get_active_hand() && Adjacent(user)))
  // Side note: don't use "usr" when you should be using "user"
  return
```
This becomes particularly important when you want to do *more* than just return, because you'd end up having to add `else` statements at specific indentation levels way below the rest of the code, which is a readability nightmare. Instead, you'll have something like...
```
if(!enoughcrayons)
  to_chat(user, "oh no you require additional crayons")
  return
```
Some of your failure states - such as the first check for having enough crayons - should have a message like that to explain what actually went wrong. Obvious stuff (canceling, moving away from the box, etc) doesn't, of course.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: I applied the host of suggested alterations to the msot recent commit with the exception of employing the W.storage_slots option because during compile it kept throwing an "undefined var" error, so I went with < 6 presuming that rayon box storage slots aren't going to change in the near or distant future.

I also applied a failure message for moving too far away in the interest of being thorough.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm (hunk 2 above)
Comment (Fox-McCloud) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: I'd go with `cardboard_box.dmi` because there's regular boxes---and if those ever get split off (not unlikely), this could cause confusion, one day.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 13]
```diff
@@ -62,3 +64,32 @@
 			for(var/mob/M in viewers(src))
 				M.show_message("<span class='notice'>\The [src] has been cut apart by [user] with \the [WC].</span>", 3, "You hear cutting.", 2)
 			qdel(src)
+			return
+		if(istype(W, /obj/item/weapon/storage/fancy/crayons))
```
Comment (Fox-McCloud) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: This is kinda weird. Just make it a `pen`; this is the consistent behavior for reskinning things.

File: code/game/objects/structures/crates_lockers/closets/cardboardbox.dm [hunk 14]
```diff
@@ -62,3 +64,32 @@
...
+			if(W.contents.len < 6)
+				to_chat(user, "You must have a full box of crayons to perform this action")
+				return
+			var/decalselection = input("Please select a decal") as null|anything in list("Atmospherics", "Bartender", "Barber", "Blueshield",	"Brig Physician", "Captain",
+				"Cargo", "Chief Engineer",	"Chaplain",	"Chef", "Chemist", "Civilian", "Clown", "CMO", "Coroner", "Detective", "Engineering", "Genetics", "HOP",
+				"HOS", "Hydroponics", "Internal Affairs Agent", "Janitor",	"Magistrate", "Mechanic", "Medical", "Mime", "Mining", "NT Representative", "Paramedic", "Pod Pilot",
+				"Prisoner",	"Research Director", "Security", "Syndicate", "Therapist", "Virology", "Warden", "Xenobiology")
+			if(!decalselection)
+				return
+			if(user.incapacitated())
```
Comment (Fox-McCloud) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: This should be the first thing checked for.

Comment (FlimFlamm) [inline code/game/objects/structures/crates_lockers/closets/cardboardbox.dm]: are you referring to the if(user.incapacitated()) or are you referring to the entire attackby() proc?

**Comment by FlimFlamm** [comment]
As of this most recent commit, Fox's suggestion to make the pen be the resprite object has been applied, although I'm not sure if i interpreted his second suggestion correctly ("This should be the first thing checked for."). I assumed he was pointing at the incapacitation check, hopefully this is correct and I've moved it to the correct place.

**Comment by FlimFlamm** [comment]
This most recent commit changes .dmi file name from 'boxes' to 'cardboard_boxes' along with updating the relevant sprite paths in cardboard_boxes.dm

**Comment by LightFire53** [comment]
I recall that someone tried to do a box species, much like the shadow people in that they need their box to survive or be burned in the light. These boxes look a lot like those ones.

**Comment by FlimFlamm** [comment]
@Lightfire53 they are in fact one and the same! Twas I who made that post so long ago https://nanotrasen.se/phpBB3/viewtopic.php?f=12&t=7740 . I only just now got around to finding out how to implement them! Cheers!
